File: 2022/js/days/day05/day05.js
import { Console, time } from 'console';
import * as fs from 'fs';

console.time('Time of part 1 and 2');

// Part 1
const array = fs.readFileSync('days/day05/input.txt', 'utf-8').split('\n');
let piles = new Array();
let piles2 = new Array();

for (let i = 0; i < 8; i++) {
    piles.push('');
}
piles.push('');

array.map((value, index) => {
    if (index < 8) {
        let ligne = value.split('');
        ligne.map((letter, i) => {
            if (i % 4 == 1 && letter != ' ') {
                piles[(i - 1) / 4] = letter + piles[(i - 1) / 4];
            }
        });
        if (index == 7) {
            piles2 = structuredClone(piles);
        }
    } else if (index >= 10) {
        let command = value.split(' ').map(Number);

        let toMove = piles[command[3] - 1].slice(
            piles[command[3] - 1].length - command[1],
            piles[command[3] - 1].length
        );

        let toMoveReversed = toMove.split('').reverse().join('');

        piles[command[3] - 1] = piles[command[3] - 1].slice(0, piles[command[3] - 1].length - command[1]);
        piles[command[5] - 1] = piles[command[5] - 1] + toMoveReversed;

        let toMove2 = piles2[command[3] - 1].slice(
            piles2[command[3] - 1].length - command[1],
            piles2[command[3] - 1].length
        );

        piles2[command[3] - 1] = piles2[command[3] - 1].slice(0, piles2[command[3] - 1].length - command[1]);
        piles2[command[5] - 1] = piles2[command[5] - 1] + toMove2;
    }
});

let giveAnswer = (tab) => {
    let answer = '';
    tab.map((val) => (answer += val[val.length - 1]));
    return answer;
};

console.timeEnd('Time of part 1 and 2');

console.log('Part 1 : ' + giveAnswer(piles));
console.log('Part 2 : ' + giveAnswer(piles2));

File: 2019/days/day05/day05.py
input_list = list(map(int, open("days/day05/input.txt", "r").read().split(",")))

op = {"1" : {"fun" : lambda x, y : x + y, "n" : 4},
      "2" : {"fun" : lambda x, y: x * y, "n" : 4} ,
      "3" : {"n" : 2},
      "4" : {"n" : 2},
      "5" : {"n" : 3},
      "6" : {"n" : 3},
      "7" : {"n" : 4},
      "8" : {"n" : 4}}

# Part 1
current = 0
output = []
while input_list[current] != 99:
    full_opcode = str(input_list[current])
    full_opcode = "0" * (5 - len(full_opcode)) + full_opcode
    opcode = full_opcode[-1]

    if opcode == "1" or opcode == "2":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        result = op[opcode]["fun"](x, y)
        input_list[input_list[current + 3]] = result
    elif opcode == "3":
        input_list[input_list[current + 1]] = int(input())
    else:
        output.append(input_list[input_list[current+1]])
    
    current += op[opcode]["n"]

print(output)

# Part 2

input_list = list(map(int, open("days/day05/input.txt", "r").read().split(",")))
current = 0
output = []
while input_list[current] != 99:
    full_opcode = str(input_list[current])
    full_opcode = "0" * (5 - len(full_opcode)) + full_opcode
    opcode = full_opcode[-1]

    if opcode == "1" or opcode == "2":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        result = op[opcode]["fun"](x, y)
        input_list[input_list[current + 3]] = result
    elif opcode == "3":
        input_list[input_list[current + 1]] = int(input())
    elif opcode == "4":
        output.append(input_list[input_list[current+1]])
    elif opcode == "5":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        if x != 0:
            current = y
            continue
    elif opcode == "6":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        if x == 0:
            current = y
            continue
    elif opcode == "7":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        input_list[input_list[current + 3]] = 1 if x < y else 0
    elif opcode == "8":
        x = int(input_list[current + 1]) if full_opcode[-3] == "1" else int(input_list[input_list[current + 1]])
        y = int(input_list[current + 2] if full_opcode[-4] == "1" else int(input_list[input_list[current + 2]]))
        input_list[input_list[current + 3]] = 1 if x == y else 0

    current += op[opcode]["n"]

print(output)